#146 added stateRegion to ContactData as an option for Addresses

DevonJSmith opened the pull request:
**Summary**

<!-- Summarize your pull request in a couple of words -->
Adds `stateRegion` as an attribute to `ContactData.cs` for use in generating meCard/vCard payloads that include the State or Region values for Address.

Additionally, reorganized `houseNumber` and `street` to more closely follow the `<houseNumber> <street>` convention for addresses.

This PR fixes/implements the following **bugs/features**:

* [x] Feature #146 (vCard State)
* [x] Bug - `houseNumber` and `street` backwards

<!-- You can skip this if you're fixing a typo or other minor things -->

What existing problem does the pull request solve?
Adds State or Region as an option when generating a Payload for vCard/meCard
Fixes issue with House Number and Street being backwards in the payload
<!-- Example: When "Adding a function to do X", explain why it is necessary to have a way to do X. -->

**Test plan**
*vCard*
Using this code to generate a payload
```csharp
 var vCard = new ContactData(ContactData.ContactOutputType.VCard3, "John", "Smith", "Johnny", "1234567890", 
                "1234567890", "1234567890", "johnny497@email.com", DateTime.Now, "www.google.com", "Pennsylvania Avenue NW", "1600",
                "Washington", "20500", "DC", "US", "this is a test");
```

Generates the following payload:
```
BEGIN:VCARD
VERSION:3.0
N:Smith;John;;;
FN:John Smith
TEL;TYPE=HOME,VOICE:1234567890
TEL;TYPE=HOME,CELL:1234567890
TEL;TYPE=WORK,VOICE:1234567890
ADR;TYPE=HOME,PREF:;;1600 Pennsylvania Avenue NW ;Washington;20500;DC;US
BDAY:20181028
URL:www.google.com
EMAIL:johnny497@email.com
NOTE:this is a test
NICKNAME:Johnny
END:VCARD
```

*meCard*
Using this call:
```csharp
var meCard = new ContactData(ContactData.ContactOutputType.MeCard, "John", "Smith", "Johnny", "1234567890", 
                "1234567890", "1234567890", "johnny497@email.com", DateTime.Now, "www.google.com", "Pennsylvania Avenue NW", "1600",
                "Washington", "20500", "DC", "US", "this is a test");
```
Generates this payload:
```
MECARD+
N:Smith, John
TEL:1234567890
TEL:1234567890
TEL:1234567890
EMAIL:johnny497@email.com
NOTE:this is a test
BDAY:20181028
ADR:,,1600 Pennsylvania Avenue NW ,Washington,20500,DC,US
URL:www.google.com
NICKNAME:Johnny
```

Importing the vCard case into Google contacts:
![capture](https://user-images.githubusercontent.com/29110132/47621931-fbf72480-dad4-11e8-87ce-01f8940ab924.PNG)

**Closing issues**

<!-- Put `closes #XXXX` (XXXX=issue number) in your comment to auto-close the issue that your PR fixes (if such). -->
`closes #146 `

codebude commented:
Hi @DevonJSmith,

thanks for you help. I really like to merge the state/region change, but before I do so, let's talk about the second change (street-housenumber) order. I get your point in re-ordering both values, but this makes only sense for American users.

In Germany (my homeland) as in most other European countries the housenumber is written after the street's name. e.g.

> John Doe
> Königsallee 1
> 40210 Düsseldorf
> Germany

So I see there no fixed solution that works for all. What do you think about adding an additional switch/flag (either _bool_ or _enum_) parameter to the method, which allows the users to choose for European/American address type? Would you mind adding it to this PR?

Ruffio commented:
IMHO, I think that implementing an option for region could be a good thing, as long as it is optional to use. Region is not used everywhere.

Regarding the street number and address sequnce, I agree with @codebude. European countries uses the notation of streetname and then housenumber. Perhaps an option to choose wich notion to use could be a good idea? 

No where in the vCard spec. does it say in which order this MUST be: https://tools.ietf.org/html/rfc6350

DevonJSmith commented:
@codebude Thank you for the feedback! I apologize for not considering the European address formats when making this change. Obviously the research I did into it skimmed over this fact (it may be possible to mail to an address from North America using the [number] [streetname] format, but that is not necessarily what is used within each European country).

Additionally, I see from my research and your example that the ZipCode and City are also transposed. 

I think that an `enum` to specify the address format would be sufficient. I'll work some more on this PR.

@Ruffio , the Region variable is indeed optional, but I do apologize for my ignorance in asserting there was a singular address convention. I will amend my PR to reflect the dynamic nature of address formats.

DevonJSmith commented:
@codebude if I do add this `enum` switch to specify European/American address format, this parameter will  likely need to be optional (in order to be consistent with the rest of the address fields.) If this value is optional, which format should be "default"?

I don't want to be presumptuous and assume that the NA address format should be default, but I suppose it depends on what the expected behavior is among the intended audience of this project. I'm interested in hearing yours and others opinions on this.

DevonJSmith commented:
I've amended the PR. I've added an `AddressOrder` enum to switch between address orders as discussed above. Here's a few examples of how these calls might work (I've also moved `stateRegion` to the end of the constructor to avoid impacting existing function calls).

### Default Order
```csharp
var defaultPayload = new ContactData(ContactData.ContactOutputType.VCard3, "John", "Smith", 
                "Johnny", "1234567890", "1234567890", "1234567890", "johns@test.com", DateTime.Now, "www.google.com", 
                "Pennsylvania Ave NW", "1600", "Washington", "20500", "US", "This is a note", "DC", ContactData.AddressOrder.Default).ToString();
```
#### Output
`BEGIN:VCARD\r\nVERSION:3.0\r\nN:Smith;John;;;\r\nFN:John Smith\r\nTEL;TYPE=HOME,VOICE:1234567890\r\nTEL;TYPE=HOME,CELL:1234567890\r\nTEL;TYPE=WORK,VOICE:1234567890\r\nADR;TYPE=HOME,PREF:;;Pennsylvania Ave NW  1600;20500;Washington;DC;US\r\nBDAY:20181111\r\nURL:www.google.com\r\nEMAIL:johns@test.com\r\nNOTE:This is a note\r\nNICKNAME:Johnny\r\nEND:VCARD`

### Reversed Order
```csharp
var reversedPayload = new ContactData(ContactData.ContactOutputType.VCard3, "John", "Smith",
                "Johnny", "1234567890", "1234567890", "1234567890", "johns@test.com", DateTime.Now, "www.google.com",
                "Pennsylvania Ave NW", "1600", "Washington", "20500", "US", "This is a note", "DC", ContactData.AddressOrder.Reversed).ToString();
```
#### Output
`BEGIN:VCARD\r\nVERSION:3.0\r\nN:Smith;John;;;\r\nFN:John Smith\r\nTEL;TYPE=HOME,VOICE:1234567890\r\nTEL;TYPE=HOME,CELL:1234567890\r\nTEL;TYPE=WORK,VOICE:1234567890\r\nADR;TYPE=HOME,PREF:;;1600 Pennsylvania Ave NW ;Washington;DC;20500;US\r\nBDAY:20181111\r\nURL:www.google.com\r\nEMAIL:johns@test.com\r\nNOTE:This is a note\r\nNICKNAME:Johnny\r\nEND:VCARD`

If this PR is accepted, we also need to make sure to update the corresponding documentation over at the Wiki (I would be happy to take care of that if needed).

Please let me know if any additional adjustments are needed.

codebude commented:
This looks really good. Thanks for your help! I'll merge your changes now.
Regarding the documentation - feel free to update it. If you won't find time for it, let me know. Latest next weekend I'll update it on myself.

codebude commented:
@DevonJSmith - I think this is the only place in the Wiki, where changes have to be done: https://github.com/codebude/QRCoder/wiki/Advanced-usage---Payload-generators#35-contactdata-mecardvcard

DevonJSmith commented:
@codebude I have added the two new parameters to that section. Please let me know if you would like any additions or further clarification!
